GC Adapter Implementation

ameerj opened the pull request:
This is exciting. This is a libusb implementation for the First-party Nintendo GameCube adapter. There is inspiration from the Dolphin team implementation, I tried not to stray too far to ensure functionality. This is also my first contribution to an established C++ project, so I apologize if my code is sloppy.

The current implementation works as another input device. Mapping works just as usual. And all 4 ports work as expected. I played for about a half hour with my brother using two controllers and there was no issue.

Currently rumble, analog triggers, and the hot plug feature are not implemented.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/yuzu-emu/yuzu?pullRequest=4137) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/yuzu-emu/yuzu?pullRequest=4137) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/yuzu-emu/yuzu?pullRequest=4137) it.</sub>

Morph1984 commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,350 @@
+// Copyright 2014 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+//*
```
Remove the //*

Morph1984 commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,116 @@
+#pragma once
```
Add the copyright header for Dolphin
```
// Copyright 2014 Dolphin Emulator Project
// Licensed under GPLv2+
// Refer to the license.txt file included.
```

Morph1984 commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,310 @@
+#include <atomic>
```
ditto

ameerj replied:
This was written from scratch by me based on sdl_impl currently in Yuzu

Morph1984 replied:
Ah, use the yuzu Emulator Project copyright header instead for gc_poller.cpp/h
```
// Copyright 2018 yuzu Emulator Project
// Licensed under GPLv2 or any later version
// Refer to the license.txt file included.
```

Morph1984 commented on src/input_common/gcadapter/gc_poller.h:
```diff
@@ -0,0 +1,59 @@
+#pragma once
```
ditto

Morph1984 commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,350 @@
+        u8 b1 = adapter_payload[1 + (9 * port) + 1];
+        u8 b2 = adapter_payload[1 + (9 * port) + 2];
+
+        if (b1 & (1 << 0))
```
All new code must be braced

ameerj replied:
Sorry i'm not sure what you mean by that.

Morph1984 replied:
```cpp
// Prefer this
if (condition) {
    // stuff
}
// over this
if (condition)
    // stuff
```

ameerj replied:
Should be resolved now!

Morph1984 commented on src/input_common/main.h:
```diff
@@ -30,6 +32,13 @@ class MotionEmu;
+GCButtonFactory* GetGCButtons();
+GCAnalogFactory* GetGCAnalogs();
+
+
```
unnecessary newline

Morph1984 commented on src/yuzu/configuration/configure_input_player.cpp:
```diff
@@ -583,6 +615,10 @@ void ConfigureInputPlayer::HandleClick(
 
     grabKeyboard();
     grabMouse();
+    if (type == InputCommon::Polling::DeviceType::Button)
+        InputCommon::GetGCButtons()->BeginConfiguration();
+    else
+        InputCommon::GetGCAnalogs()->BeginConfiguration();
```
These should also be braced

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,375 @@
+
+namespace GCAdapter {
+
+static libusb_device_handle* usb_adapter_handle = nullptr;
```
All of this file-scope state should be encapsulated into a class. Please don't use global variables.

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,375 @@
+#include "common/threadsafe_queue.h"
+#include "input_common/gcadapter/gc_adapter.h"
+
+Common::SPSCQueue<GCPadStatus> pad_queue[4];
```
`std::array`, ditto below where applicable

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,375 @@
+}
+
+void PadToState(GCPadStatus pad, GCState& state) {
+    //std::lock_guard lock{s_mutex};
```
Commented out code

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,120 @@
+    u8 triggerRight; // 0 <= triggerRight <= 255
+    bool isConnected{true};
+
+    static const u8 MAIN_STICK_CENTER_X = 0x80;
```
These static variables can all be constexpr

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,120 @@
+#include <functional>
+#include "common/common_types.h"
+
+
```
The GC adapter namespace should start here.

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,120 @@
+
+};
+
+enum PadAxes { STICK_X, STICK_Y, SUBSTICK_X, SUBSTICK_Y, TRIGGER_LEFT, TRIGGER_RIGHT };
```
```suggestion
enum class PadAxes {
    StickX,
    StickY,
    SubstickX,
    SubstickY,
    TriggerLeft,
    TriggerRight,
};
```

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,120 @@
+
+struct GCPadStatus {
+    u16 button;      // Or-ed PAD_BUTTON_* and PAD_TRIGGER_* bits
+    u8 stickX;       // 0 <= stickX       <= 255
```
Our naming convention is `variable_name`, not `variableName`. Ditto where applicable elsewhere.

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,120 @@
+
+
+namespace GCAdapter {
+enum ControllerTypes {
```
```cpp
enum class ControllerTypes {
    None,
    Wired,
    Wireless,
};
```

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,120 @@
+};
+
+// Current adapter status: detected/not detected/in error (holds the error code)
+static int current_status = NO_ADAPTER_DETECTED;
```
This is a file-scope global variable in a header. I'd strongly advise against doing this, since it can result in hard to track down bugs, depending on inclusion order.

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,120 @@
+
+GCPadStatus CheckStatus(int port, u8 adapter_payload[37]);
+/// Initialize the GC Adapter capture and read sequence
+void Init();
```
This entire interface should be refactored into a stateful class to avoid previously mentioned global variables.

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,317 @@
+#include "common/threadsafe_queue.h"
+
+// Using extern as to avoid multply defined symbols.
+extern Common::SPSCQueue<GCPadStatus> pad_queue[4];
```
No global variables

lioncash commented:
Just to make it a top level comment so it's not lost in the review noise, the gc_adapter interface should be converted into a stateful class to eliminate the global variables, like all the other input devices are currently organized.

ameerj commented:
@lioncash Thank you so much for that amazing feedback. I think I addressed all of your concerns, and most likely introduced new concerns. I am particularly unsure of a possible memory leak. The Singleton Adapter class is a "raw" pointer, not sure if that's a bad idea.

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,163 @@
+    TriggerRight,
+    Undefined,
+};
+const struct GCPadConstants {
```
Why is a separate class used for these? Them being `static constexpr` in the class was totally fine.

ameerj replied:
I figured one single instance of these constants is better that having it included with every `GCPadStatus` object that gets created. And my attempt at declaring a variable as `constexpr` made the compiler yell at me.

lioncash replied:
Why would the compiler yell at you? It's totally valid C++ to make `static constexpr` constants in a class declaration.

lioncash commented on src/input_common/main.cpp:
```diff
@@ -2,10 +2,14 @@
 // Licensed under GPLv2 or any later version
 // Refer to the license.txt file included.
 
+#include <iostream>
```
This isn't necessary

lioncash commented on src/input_common/main.cpp:
```diff
@@ -22,8 +26,15 @@ static std::shared_ptr<MotionEmu> motion_emu;
+static std::shared_ptr<GCAnalogFactory> gcanalog;
 
 void Init() {
+    gcbuttons = std::make_shared<GCButtonFactory>();
```
Why do these need to be shared_ptr instances over unique_ptr?

ameerj replied:
Input::RegisterFactory expects the arg to be of take shared_ptr

lioncash commented on src/input_common/gcadapter/gc_poller.h:
```diff
@@ -0,0 +1,65 @@
+    void BeginConfiguration();
+    void EndConfiguration();
+
+    bool IsPolling() {
```
```suggestion
    bool IsPolling() const {
```

lioncash commented on src/input_common/gcadapter/gc_poller.h:
```diff
@@ -0,0 +1,65 @@
+    void BeginConfiguration();
+    void EndConfiguration();
+
+    bool IsPolling() {
```
```suggestion
    bool IsPolling() const {
```

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,315 @@
+}
+
+GCButton::~GCButton() {
+    // GCAdapter::Shutdown();
```
If this code isn't necessary, then this should be defaulted like so:

```cpp
GCButton::~GCButton() = default;
```

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,163 @@
+    void BeginConfiguration();
+    void EndConfiguration();
+
+    std::array<Common::SPSCQueue<GCPadStatus>, 4>& GetPadQueue();
```
There should also be const variants of these getters.

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,163 @@
+
+    int current_status = NO_ADAPTER_DETECTED;
+    libusb_device_handle* usb_adapter_handle = nullptr;
+    ControllerTypes adapter_controllers_status[4] = {ControllerTypes::None, ControllerTypes::None,
```
```cpp
std::array<ContollerTypes, 4> adapter_controllers_status{};
```

Is sufficient for initializing all of these.

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,372 @@
+        current_status = NO_ADAPTER_DETECTED;
+    }
+
+    for (int i = 0; i < 4; i++) {
```
```cpp
adapter_controllers_status.fill(ControllerTypes::None);
```

ameerj replied:
WOW!!! I needed this in my life.

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,372 @@
+        adapter_input_thread.join();
+    }
+
+    for (int i = 0; i < 4; i++) {
```
```suggestion
    adapter_controllers_status.fill(ControllerTypes::None);
```

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,372 @@
+            payload_size = payload_size_in;
+        }
+
+        GCPadStatus pad[4];
```
std::array

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,372 @@
+                                  sizeof(adapter_payload), &payload_size_in, 32);
+
+        int payload_size = 0;
+        u8 controller_payload_copy[37];
```
std::array

ameerj replied:
Unfortunately, libusb is actually a C library that's C++ compatible, expects `unsigned char*`

lioncash replied:
This can still be a std::array. You just use its `.data()` member function for passing to C libraries.

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,372 @@
+    LOG_INFO(Input, "GC Adapter Read() thread started");
+
+    int payload_size_in;
+    u8 adapter_payload[37];
```
std::array

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,163 @@
+    ADAPTER_DETECTED = 1,
+};
+
+/// Singleton Adapter class
```
Please don't use singletons, these are still global variables.

This prevents the resources from being released when emulation ends. This should only be created once and be contained by the factory class. This way whenever the factory is unregistered, then the class automatically gets deallocated.

ameerj replied:
The GC adapter can only be registered by one instance of the GCAdapter class. I'll have to think this through more but there's a separate GCAnalog and GCButton factory that both need to reference the same GCAdapter class.

lioncash replied:
A shared_ptr can be passed to both of these factories during initialization (lifetime of the adapter is shared between the two factories).

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,315 @@
+
+void GCAnalogFactory::BeginConfiguration() {
+    polling = true;
+    for (int i = 0; i < 4; i++) {
```
The adapter should likely have a function to perform this clearing itself. This way a magic size value doesn't need to be hardcoded into this class.

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,315 @@
+
+void GCButtonFactory::BeginConfiguration() {
+    polling = true;
+    for (int i = 0; i < 4; i++) {
```
Ditto

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,301 @@
+};
+
+GCButtonFactory::GCButtonFactory(std::shared_ptr<GCAdapter::Adapter> adapter_)
+    : adapter(adapter_) {}
```
```suggestion
    : adapter(std::move(adapter_)) {}
```

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,301 @@
+    explicit GCAxisButton(int port_, int axis_, float threshold_, bool trigger_if_greater_,
+                          std::shared_ptr<GCAdapter::Adapter> adapter)
+        : port(port_), axis(axis_), threshold(threshold_), trigger_if_greater(trigger_if_greater_),
+          gcadapter(adapter) {}
```
ditto

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,301 @@
+class GCButton final : public Input::ButtonDevice {
+public:
+    explicit GCButton(int port_, int button_, int axis_,
+                      std::shared_ptr<GCAdapter::Adapter> adapter)
```
```suggestion
                      GCAdapter::Adapter* adapter)
```

Given this class uses the pointer as a non-owning pointer

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,301 @@
+private:
+    const int port;
+    const int button;
+    std::shared_ptr<GCAdapter::Adapter> gcadapter;
```
Ditto

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,301 @@
+    const int axis;
+    float threshold;
+    bool trigger_if_greater;
+    std::shared_ptr<GCAdapter::Adapter> gcadapter;
```
ditto

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,301 @@
+class GCAnalog final : public Input::AnalogDevice {
+public:
+    GCAnalog(int port_, int axis_x_, int axis_y_, float deadzone_,
+             std::shared_ptr<GCAdapter::Adapter> adapter)
```
This can be a regular pointer

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,301 @@
+    const int axis_y;
+    const float deadzone;
+    mutable std::mutex mutex;
+    std::shared_ptr<GCAdapter::Adapter> gcadapter;
```
Ditto

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,301 @@
+        return (gcadapter->GetPadState()[port].axes.at(axis) - 128.0f) / 95.0f;
+    }
+
+    std::tuple<float, float> GetAnalog(int axis_x, int axis_y) const {
```
```suggestion
    std::pair<float, float> GetAnalog(int axis_x, int axis_y) const {
```

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,301 @@
+            return std::make_tuple(x / r * (r - deadzone) / (1 - deadzone),
+                                   y / r * (r - deadzone) / (1 - deadzone));
+        }
+        return std::make_tuple<float, float>(0.0f, 0.0f);
```
```suggestion
        return {0.0f, 0.0f};
```

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,301 @@
+        const auto [x, y] = GetAnalog(axis_x, axis_y);
+        const float r = std::sqrt((x * x) + (y * y));
+        if (r > deadzone) {
+            return std::make_tuple(x / r * (r - deadzone) / (1 - deadzone),
```
This can just use bracing syntax

```suggestion
            return {x / r * (r - deadzone) / (1 - deadzone),
```

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,301 @@
+
+/// An analog device factory that creates analog devices from GC Adapter
+GCAnalogFactory::GCAnalogFactory(std::shared_ptr<GCAdapter::Adapter> adapter_)
+    : adapter(adapter_) {}
```
```suggestion
    : adapter(std::move(adapter_)) {}
```

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,301 @@
+ *     - "axis_y": the index of the axis to be bind as y-axis
+ */
+std::unique_ptr<Input::AnalogDevice> GCAnalogFactory::Create(const Common::ParamPackage& params) {
+    const std::string guid = params.Get("guid", "0");
```
This is unused, and should be marked as `[[maybe_unused]]` if that's intentional.

ameerj replied:
Good catch!

lioncash commented on src/input_common/gcadapter/gc_poller.h:
```diff
@@ -0,0 +1,65 @@
+ */
+class GCButtonFactory final : public Input::Factory<Input::ButtonDevice> {
+public:
+    GCButtonFactory(std::shared_ptr<GCAdapter::Adapter> adapter_);
```
```suggestion
    explicit GCButtonFactory(std::shared_ptr<GCAdapter::Adapter> adapter_);
```

lioncash commented on src/input_common/gcadapter/gc_poller.h:
```diff
@@ -0,0 +1,65 @@
+/// An analog device factory that creates analog devices from GC Adapter
+class GCAnalogFactory final : public Input::Factory<Input::AnalogDevice> {
+public:
+    GCAnalogFactory(std::shared_ptr<GCAdapter::Adapter> adapter_);
```
```suggestion
    explicit GCAnalogFactory(std::shared_ptr<GCAdapter::Adapter> adapter_);
```

lioncash commented on src/input_common/main.cpp:
```diff
@@ -22,8 +25,17 @@ static std::shared_ptr<MotionEmu> motion_emu;
 static std::unique_ptr<SDL::State> sdl;
 #endif
 static std::unique_ptr<CemuhookUDP::State> udp;
+static std::shared_ptr<GCAdapter::Adapter> gcadapter;
```
This doesn't need to be a file-scope variable, this can just be local to `Init()`

lioncash commented on src/input_common/main.h:
```diff
@@ -7,6 +7,8 @@
 #include <memory>
 #include <string>
 #include <vector>
+#include "input_common/gcadapter/gc_adapter.h"
```
These can be replaced with forward declarations, like with Keyboard a little further down in the file.

ameerj replied:
I get a `use of undefined type 'InputCommon::GCButtonFactory'` error when I do the forward declaration. Wanted to play it safe with the header file inclusion.

lioncash commented on src/input_common/main.cpp:
```diff
@@ -22,8 +25,16 @@ static std::shared_ptr<MotionEmu> motion_emu;
+static std::shared_ptr<GCAnalogFactory> gcanalog;
 
 void Init() {
+    std::shared_ptr<GCAdapter::Adapter> gcadapter = std::make_shared<GCAdapter::Adapter>();
```
```suggestion
    auto gcadapter = std::make_shared<GCAdapter::Adapter>();
```

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,157 @@
+    std::array<GCState, 4> state;
+};
+
+} // end of namespace GCAdapter
```
```suggestion
} // namespace GCAdapter
```

ameerj commented:
Just a note, for windows the WinUSB driver for the gc adapter needs to be installed using Zadig (Just like [Dolphin's installation process](https://dolphin-emu.org/docs/guides/how-use-official-gc-controller-adapter-wii-u/#Using_Zadig))

I briefly tested the build artifact on linux and the settings detected my adapter without any installations needed!

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+    bool get_origin = false;
+
+    ControllerTypes type = ControllerTypes(adapter_payload[1 + (9 * port)] >> 4);
+    if (type != ControllerTypes::None)
+        get_origin = true;
```
```suggestion
    if (type != ControllerTypes::None) {
        get_origin = true;
    }
```

(formatting might be broken since I'm doing it by hand)

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+void Adapter::Read() {
+    LOG_INFO(Input, "GC Adapter Read() thread started");
+
+    int payload_size_in, payload_size;
```
std::pair?

ameerj replied:
I thought that might have unnecessary added overhead, since these vars remain local.

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+            LOG_ERROR(Input, "error reading payload (size: %d, type: %02x)", payload_size,
+                      controller_payload_copy[0]);
+        } else {
+            for (int port = 0; port < 4; port++) {
```
Is this the size of pad? If so:
```suggestion
            for (int port = 0; port < pad.size(); port++) {
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+                pad[port] = CheckStatus(port, controller_payload_copy);
+            }
+        }
+        for (int port = 0; port < 4; port++) {
```
Is this the size of pad? If so:
```suggestion
        for (int port = 0; port < pad.size(); port++) {
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+        u8 b1 = adapter_payload[1 + (9 * port) + 1];
+        u8 b2 = adapter_payload[1 + (9 * port) + 2];
+
+        if (b1 & (1 << 0)) {
+            pad.button |= PAD_BUTTON_A;
+        }
+        if (b1 & (1 << 1)) {
+            pad.button |= PAD_BUTTON_B;
+        }
+        if (b1 & (1 << 2)) {
+            pad.button |= PAD_BUTTON_X;
+        }
+        if (b1 & (1 << 3)) {
+            pad.button |= PAD_BUTTON_Y;
+        }
+
+        if (b1 & (1 << 4)) {
+            pad.button |= PAD_BUTTON_LEFT;
+        }
+        if (b1 & (1 << 5)) {
+            pad.button |= PAD_BUTTON_RIGHT;
+        }
+        if (b1 & (1 << 6)) {
+            pad.button |= PAD_BUTTON_DOWN;
+        }
+        if (b1 & (1 << 7)) {
+            pad.button |= PAD_BUTTON_UP;
+        }
```
This *might* be a little better (not sure if it breaks anything)
Once again formatting is broken

```suggestion
enum b1_pad_buttons : int {
PAD_BUTTON_A,
PAD_BUTTON_B,
PAD_BUTTON_X,
PAD_BUTTON_Y,
PAD_BUTTON_LEFT,
PAD_BUTTON_RIGHT,
PAD_BUTTON_DOWN,
PAD_BUTTON_UP
};

constexpr std::array<int, 8> b1_buttons {
PAD_BUTTON_A, PAD_BUTTON_B, PAD_BUTTON_X, PAD_BUTTON_Y,
PAD_BUTTON_LEFT, PAD_BUTTON_RIGHT, PAD_BUTTON_DOWN, PAD_BUTTON_UP
};

for(int j = 0; j < b1_buttons.size(); j++) {
        if (b1 & (1 << j)) {
            pad.button |= b1_buttons[j];
        }
}
```

ameerj replied:
I was debating whether to keep these verbose, or to do a trick like you are proposing. gc_poller.cpp line 94 has a similar approach. I may be wrong but either approach probably won't have any performance difference. 

> 
> 
> This _might_ be a little better (not sure if it breaks anything)

I was debating whether to keep these verbose, or to do a trick like you are proposing. gc_poller.cpp line 94 has a similar approach. I may be wrong but either approach probably won't have any performance difference.

ameerj replied:
@lioncash do you have a preference for which approach we should move forward with?

ameerj replied:
I addressed this in the latest commit. LMK what you think.

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+            pad.button |= PAD_BUTTON_UP;
+        }
+
+        if (b2 & (1 << 0)) {
+            pad.button |= PAD_BUTTON_START;
+        }
+        if (b2 & (1 << 1)) {
+            pad.button |= PAD_TRIGGER_Z;
+        }
+        if (b2 & (1 << 2)) {
+            pad.button |= PAD_TRIGGER_R;
+        }
+        if (b2 & (1 << 3)) {
+            pad.button |= PAD_TRIGGER_L;
+        }
```
If the above solution is better and is accepted, same as above but s/b1/b2/

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+
+    libusb_device** devs; // pointer to list of connected usb devices
+
+    int cnt = libusb_get_device_list(libusb_ctx, &devs); // get the list of devices
```
```suggestion
    const int cnt = libusb_get_device_list(libusb_ctx, &devs); // get the list of devices
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+    adapter_controllers_status[port] = type;
+
+    if (adapter_controllers_status[port] != ControllerTypes::None) {
+        u8 b1 = adapter_payload[1 + (9 * port) + 1];
```
```suggestion
        const u8 b1 = adapter_payload[1 + (9 * port) + 1];
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+
+    if (adapter_controllers_status[port] != ControllerTypes::None) {
+        u8 b1 = adapter_payload[1 + (9 * port) + 1];
+        u8 b2 = adapter_payload[1 + (9 * port) + 2];
```
```suggestion
        const u8 b2 = adapter_payload[1 + (9 * port) + 2];
```

VolcaEM commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,299 @@
+GCButton::~GCButton() = default;
+
+std::unique_ptr<Input::ButtonDevice> GCButtonFactory::Create(const Common::ParamPackage& params) {
+    int button_id = params.Get("button", 0);
```
```suggestion
    const int button_id = params.Get("button", 0);
```

VolcaEM commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,299 @@
+
+std::unique_ptr<Input::ButtonDevice> GCButtonFactory::Create(const Common::ParamPackage& params) {
+    int button_id = params.Get("button", 0);
+    int port = params.Get("port", 0);
```
```suggestion
    const int port = params.Get("port", 0);
```

VolcaEM commented on src/input_common/keyboard.cpp:
```diff
@@ -49,8 +49,10 @@ class KeyButtonList {
-            if (pair.key_code == key_code)
+            if (pair.key_code == key_code) {
                 pair.key_button->status.store(pressed);
+                break;
```
Has this been tested properly to not break anything?

ameerj replied:
Forgot about this, it triggered my OCD when I was exploring the internals. It looped through ~600 keys only to change the status of one, then continued through the rest.

Will revert back now that I think about it, it's needed if multiple keys are mapped to the same key_code.

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+
+bool Adapter::CheckDeviceAccess(libusb_device* device) {
+    libusb_device_descriptor desc;
+    int ret = libusb_get_device_descriptor(device, &desc);
+    if (ret) {
```
```suggestion
    const int got_device_descriptor = libusb_get_device_descriptor(device, &desc);
    if (got_device_descriptor) {
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+        // This isn�t the device we are looking for.
+        return false;
+    }
+    ret = libusb_open(device, &usb_adapter_handle);
+
+    if (ret == LIBUSB_ERROR_ACCESS) {
```
```suggestion
    const int could_open = libusb_open(device, &usb_adapter_handle);

    if (could_open == LIBUSB_ERROR_ACCESS) {
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+                  desc.idProduct);
+        return false;
+    }
+    if (ret) {
```
```suggestion
    if (could_open) {
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+        return false;
+    }
+    if (ret) {
+        LOG_ERROR(Input, "libusb_open failed to open device with error = %d", ret);
```
```suggestion
        LOG_ERROR(Input, "libusb_open failed to open device with error = %d", could_open);
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+    int ret = libusb_get_device_descriptor(device, &desc);
+    if (ret) {
+        // could not acquire the descriptor, no point in trying to use it.
+        LOG_ERROR(Input, "libusb_get_device_descriptor failed with error: %d", ret);
```
```suggestion
        LOG_ERROR(Input, "libusb_get_device_descriptor failed with error: %d", got_device_descriptor);
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+        return false;
+    }
+
+    ret = libusb_kernel_driver_active(usb_adapter_handle, 0);
```
```suggestion
    int ret = libusb_kernel_driver_active(usb_adapter_handle, 0);
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+        return false;
+    }
+
+    ret = libusb_claim_interface(usb_adapter_handle, 0);
```
```suggestion
    const int interface_claimed = libusb_claim_interface(usb_adapter_handle, 0);
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+    }
+
+    ret = libusb_claim_interface(usb_adapter_handle, 0);
+    if (ret) {
```
```suggestion
    if (interface_claimed) {
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+
+    ret = libusb_claim_interface(usb_adapter_handle, 0);
+    if (ret) {
+        LOG_ERROR(Input, "libusb_claim_interface failed with error = %d", ret);
```
```suggestion
        LOG_ERROR(Input, "libusb_claim_interface failed with error = %d", interface_claimed);
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,371 @@
+    if (ret == 1) {
+        ret = libusb_detach_kernel_driver(usb_adapter_handle, 0);
+        if (ret != 0 && ret != LIBUSB_ERROR_NOT_SUPPORTED) {
+            LOG_ERROR(Input, "libusb_detach_kernel_driver failed with error = %d", ret);
```
Why no return here?

ameerj replied:
This rolls over and is handled by the next block that closes the handle and returns. This logic came from Dolphin and I couldn't do any better to clean it up.

ogniK5377 commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,162 @@
+    PAD_ERR_STATUS = 0x8000,
+};
+
+enum PadButton {
```
```suggestion
enum class PadButton {
```

ogniK5377 commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,162 @@
+    Undefined,
+};
+
+struct GCPadStatus {
+    u16 button;       // Or-ed PAD_BUTTON_* and PAD_TRIGGER_* bits
+    u8 stick_x;       // 0 <= stick_x       <= 255
+    u8 stick_y;       // 0 <= stick_y       <= 255
+    u8 substick_x;    // 0 <= substick_x    <= 255
+    u8 substick_y;    // 0 <= substick_y    <= 255
+    u8 trigger_left;  // 0 <= trigger_left  <= 255
+    u8 trigger_right; // 0 <= trigger_right <= 255
+
+    static constexpr u8 MAIN_STICK_CENTER_X = 0x80;
+    static constexpr u8 MAIN_STICK_CENTER_Y = 0x80;
+    static constexpr u8 MAIN_STICK_RADIUS = 0x7f;
+    static constexpr u8 C_STICK_CENTER_X = 0x80;
+    static constexpr u8 C_STICK_CENTER_Y = 0x80;
+    static constexpr u8 C_STICK_RADIUS = 0x7f;
+    static constexpr u8 TRIGGER_CENTER = 20;
+    static constexpr u8 THRESHOLD = 10;
+
+    u8 port;
+    PadAxes axis = PadAxes::Undefined;
+    u8 axis_value = 255;
```
```suggestion
struct GCPadStatus {
    u16 button{};       // Or-ed PAD_BUTTON_* and PAD_TRIGGER_* bits
    u8 stick_x{};       // 0 <= stick_x       <= 255
    u8 stick_y{};       // 0 <= stick_y       <= 255
    u8 substick_x{};    // 0 <= substick_x    <= 255
    u8 substick_y{};    // 0 <= substick_y    <= 255
    u8 trigger_left{};  // 0 <= trigger_left  <= 255
    u8 trigger_right{}; // 0 <= trigger_right <= 255

    static constexpr u8 MAIN_STICK_CENTER_X = 0x80;
    static constexpr u8 MAIN_STICK_CENTER_Y = 0x80;
    static constexpr u8 MAIN_STICK_RADIUS = 0x7f;
    static constexpr u8 C_STICK_CENTER_X = 0x80;
    static constexpr u8 C_STICK_CENTER_Y = 0x80;
    static constexpr u8 C_STICK_RADIUS = 0x7f;
    static constexpr u8 TRIGGER_CENTER = 20;
    static constexpr u8 THRESHOLD = 10;

    u8 port{};
    PadAxes axis{PadAxes::Undefined};
    u8 axis_value{255};
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -70,18 +72,11 @@ GCPadStatus Adapter::GetPadStatus(int port, const std::array<u8, 37>& adapter_pa
-    state.buttons.insert_or_assign(PAD_TRIGGER_Z, pad.button & PAD_TRIGGER_Z);
-    state.buttons.insert_or_assign(PAD_TRIGGER_L, pad.button & PAD_TRIGGER_L);
-    state.buttons.insert_or_assign(PAD_TRIGGER_R, pad.button & PAD_TRIGGER_R);
+    for (auto button : PadButtonArray) {
```
```suggestion
    for (auto& button : PadButtonArray) {
```
I think

ameerj replied:
Both worked. I'm sure const reference is preferred. Thanks for pointing that out.

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,349 @@
+            }
+        }
+
+        for (int j = 0; j < b2_buttons.size(); j++) {
```
You can use i here instead of j if you want

ameerj replied:
I've had bugs where I overlooked the index variable and rewrote functions before figuring out what the bug was. Did that just as a sanity check tbh. i for b1, j for b2. :)

Morph1984 commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,349 @@
+}
+
+void Adapter::PadToState(const GCPadStatus& pad, GCState& state) {
+    for (auto const& button : PadButtonArray) {
```
our convention is left const, so
```suggestion
    for (const auto& button : PadButtonArray) {
```

Morph1984 commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,349 @@
+    }
+
+    if (desc.idVendor != 0x057e || desc.idProduct != 0x0337) {
+        // This isnt the device we are looking for.
```
```suggestion
        // This isn't the device we are looking for.
```

Morph1984 commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,349 @@
+
+    libusb_device** devs; // pointer to list of connected usb devices
+
+    const int cnt = libusb_get_device_list(libusb_ctx, &devs); // get the list of devices
```
This causes conversion warnings
```
src\input_common\gcadapter\gc_adapter.cpp(194,43): warning C4244: 'initializing': conversion from 'ssize_t' to 'int', possible loss of data
src\input_common\gcadapter\gc_adapter.cpp(194,19): warning C4244: 'initializing': conversion from 'ssize_t' to 'const int', possible loss of data
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,349 @@
+            payload_size_copy = payload_size_in;
+        }
+
+        if (payload_size_copy != sizeof(adapter_payload_copy) ||
```
You may want to add an assert for this (not sure)

ameerj replied:
Will an assert stop execution of yuzu?

VolcaEM replied:
Not in mainline/EA thanks to https://github.com/yuzu-emu/yuzu/pull/1340, but thinking about it it's best to *not* add an assert here

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,349 @@
+            adapter_payload_copy[0] != LIBUSB_DT_HID) {
+            LOG_ERROR(Input, "error reading payload (size: %d, type: %02x)", payload_size_copy,
+                      adapter_payload_copy[0]);
+            adapter_thread_running = false; // error reading from adapter, stop reading.
```
Is there a particular reason for not returning here?
That way, we can remove the else statement

ameerj replied:
For now, you're right. Might need to revisit for the hotplug implementation.

Morph1984 commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,349 @@
+        const u8 b1 = adapter_payload[1 + (9 * port) + 1];
+        const u8 b2 = adapter_payload[1 + (9 * port) + 2];
+
+        for (int i = 0; i < b1_buttons.size(); i++) {
```
```suggestion
        for (std::size_t i = 0; i < b1_buttons.size(); ++i) {
```

Morph1984 commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,349 @@
+            }
+        }
+
+        for (int j = 0; j < b2_buttons.size(); j++) {
```
```suggestion
        for (std::size_t j = 0; j < b2_buttons.size(); ++j) {
```

Morph1984 commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,349 @@
+                      adapter_payload_copy[0]);
+            adapter_thread_running = false; // error reading from adapter, stop reading.
+        } else {
+            for (int port = 0; port < pads.size(); port++) {
```
```suggestion
            for (std::size_t port = 0; port < pads.size(); ++port) {
```

Morph1984 commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,349 @@
+                pads[port] = GetPadStatus(port, adapter_payload_copy);
+            }
+        }
+        for (int port = 0; port < pads.size(); port++) {
```
```suggestion
        for (std::size_t port = 0; port < pads.size(); ++port) {
```

VolcaEM commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,349 @@
+    }
+
+    detect_thread_running = true;
+    detect_thread = std::thread([=] { ScanThreadFunc(); });
```
I think you can just do
```suggestion
    detect_thread = std::thread(&Adapter::ScanThreadFunc);
```

ameerj replied:
Unfortunately not, for member methods I needed to do that weird way since the instance itself is calling its own method, it's not a static public method.

VolcaEM commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,262 @@
+    const int button_id = params.Get("button", 0);
+    const int port = params.Get("port", 0);
+    // For Axis buttons, used by the binary sticks.
+    if (params.Has("axis")) {
+        const int axis = params.Get("axis", 0);
+        const float threshold = params.Get("threshold", 0.5f);
+        const std::string direction_name = params.Get("direction", "");
+        bool trigger_if_greater;
+        if (direction_name == "+") {
+            trigger_if_greater = true;
+        } else if (direction_name == "-") {
+            trigger_if_greater = false;
+        } else {
+            trigger_if_greater = true;
+            LOG_ERROR(Input, "Unknown direction {}", direction_name);
+        }
+        return std::make_unique<GCAxisButton>(port, axis, threshold, trigger_if_greater,
+                                              adapter.get());
+    }
+
+    std::unique_ptr<GCButton> button =
+        std::make_unique<GCButton>(port, button_id, params.Get("axis", 0), adapter.get());
+    return std::move(button);
```
(My) formatting is broken
```suggestion
    if (!params.Has("axis")) {
            std::unique_ptr<GCButton> button =
        std::make_unique<GCButton>(port, button_id, params.Get("axis", 0), adapter.get());
    return std::move(button);
    }
    
    const int axis = params.Get("axis", 0);
        const float threshold = params.Get("threshold", 0.5f);
        const std::string direction_name = params.Get("direction", "");
        bool trigger_if_greater;
        if (direction_name == "+") {
            trigger_if_greater = true;
        } else if (direction_name == "-") {
            trigger_if_greater = false;
        } else {
            trigger_if_greater = true;
            LOG_ERROR(Input, "Unknown direction {}", direction_name);
        }
        return std::make_unique<GCAxisButton>(port, axis, threshold, trigger_if_greater,
                                              adapter.get());
```

ameerj replied:
How does flipping the branch order necessarily help?

VolcaEM replied:
Readability

ameerj replied:
I did you one better !

VolcaEM commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,262 @@
+            // or to use a while loop shifting the bits to test and set the value.
+
+            for (auto button : GCAdapter::PadButtonArray) {
+                u16 button_value = static_cast<u16>(button);
```
I think this can be const
```suggestion
                const u16 button_value = static_cast<u16>(button);
```

Morph1984 commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,349 @@
+
+    adapter_controllers_status.fill(ControllerTypes::None);
+
+    libusb_device** devs; // pointer to list of connected usb devices
+
+    const std::size_t cnt = libusb_get_device_list(libusb_ctx, &devs); // get the list of devices
+
+    for (int i = 0; i < cnt; i++) {
+        if (CheckDeviceAccess(devs[i])) {
+            // GC Adapter found and accessible, registering it
+            GetGCEndpoint(devs[i]);
```
```suggestion
    libusb_device** devices; // pointer to list of connected usb devices

    const std::size_t device_count = libusb_get_device_list(libusb_ctx, &devices); // get the list of devices

    for (std::size_t index = 0; index < device_count; ++index) {
        if (CheckDeviceAccess(devices[index])) {
            // GC Adapter found and accessible, registering it
            GetGCEndpoint(devices[index]);
```

Morph1984 commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -0,0 +1,262 @@
+    Common::ParamPackage params;
+    GCAdapter::GCPadStatus pad;
+    auto& queue = adapter->GetPadQueue();
+    for (int port = 0; port < queue.size(); port++) {
```
```suggestion
    for (std::size_t port = 0; port < queue.size(); ++port) {
```

ogniK5377 commented:
Build is failing on linux

ameerj commented:
Windows does not properly support the libusb hotplug feature. Dolphin resolved this by having a scan thread constantly scan and update the status of the adapter every 500 ms on Windows. 
I am unsure if we want to try to replicate this functionality, perhaps in a future PR? After my latest commit, if the device is unplugged while yuzu is running, yuzu needs to be shut down and rebooted before it can recognize the adapter again. and the adapter must be plugged in before yuzu is run.

FearlessTobi commented on src/input_common/sdl/sdl_impl.cpp:
```diff
@@ -444,6 +444,7 @@ class SDLButtonFactory final : public Input::Factory<Input::ButtonDevice> {
 class SDLAnalogFactory final : public Input::Factory<Input::AnalogDevice> {
 public:
     explicit SDLAnalogFactory(SDLState& state_) : state(state_) {}
+
```
unnecessary diff

FearlessTobi commented on src/input_common/main.cpp:
```diff
@@ -88,7 +111,6 @@ std::vector<std::unique_ptr<DevicePoller>> GetPollers(DeviceType type) {
 #ifdef HAVE_SDL2
     pollers = sdl->GetPollers(type);
 #endif
-
```
unnecessary diff

FearlessTobi commented on src/input_common/main.cpp:
```diff
@@ -34,7 +45,6 @@ void Init() {
 #ifdef HAVE_SDL2
     sdl = SDL::Init();
 #endif
-
```
unnecessary diff

FearlessTobi commented on src/input_common/keyboard.cpp:
```diff
@@ -49,8 +49,9 @@ class KeyButtonList {
     void ChangeKeyStatus(int key_code, bool pressed) {
         std::lock_guard guard{mutex};
         for (const KeyButtonPair& pair : list) {
-            if (pair.key_code == key_code)
```
unnecessary diff

bunnei commented:
re-tagging, as this seems fine now

bunnei commented:
@ameerj sorry! Needed to untag again, as this conflicts with another EA-tagged PR, and was stalling the build pipeline. Let me investigate more and see what the issue is :(

ameerj commented:
> 
> 
> @ameerj sorry! Needed to untag again, as this conflicts with another EA-tagged PR, and was stalling the build pipeline. Let me investigate more and see what the issue is :(

@bunnei no worries! let me know if there's anything I need to do.

lioncash commented on src/input_common/gcadapter/gc_poller.cpp:
```diff
@@ -60,10 +66,20 @@ GCButton::~GCButton() = default;
+
+    // button is not an axis/stick button
+    if (button_id != PAD_STICK_ID) {
+        std::unique_ptr<GCButton> button =
```
```suggestion
        auto button =
```

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,369 @@
+        const u8 b2 = adapter_payload[1 + (9 * port) + 2];
+
+        for (std::size_t i = 0; i < b1_buttons.size(); ++i) {
+            if (b1 & (1 << i)) {
```
```suggestion
            if ((b1 & (1U << i)) != 0) {
```

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,369 @@
+        }
+
+        for (std::size_t j = 0; j < b2_buttons.size(); ++j) {
+            if (b2 & (1 << j)) {
```
```suggestion
            if ((b2 & (1U << j)) != 0) {
```

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,369 @@
+
+    adapter_controllers_status[port] = type;
+
+    constexpr std::array<PadButton, 8> b1_buttons{
```
```suggestion
    static constexpr std::array<PadButton, 8> b1_buttons{
```

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,369 @@
+    constexpr std::array<PadButton, 8> b1_buttons{
+        PadButton::PAD_BUTTON_A,    PadButton::PAD_BUTTON_B,    PadButton::PAD_BUTTON_X,
+        PadButton::PAD_BUTTON_Y,    PadButton::PAD_BUTTON_LEFT, PadButton::PAD_BUTTON_RIGHT,
+        PadButton::PAD_BUTTON_DOWN, PadButton::PAD_BUTTON_UP};
```
A comma should be after the last element so that the formatting is nicer.

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,369 @@
+        PadButton::PAD_BUTTON_Y,    PadButton::PAD_BUTTON_LEFT, PadButton::PAD_BUTTON_RIGHT,
+        PadButton::PAD_BUTTON_DOWN, PadButton::PAD_BUTTON_UP};
+
+    constexpr std::array<PadButton, 4> b2_buttons{
```
```suggestion
    static constexpr std::array<PadButton, 4> b2_buttons{
```

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,369 @@
+
+    constexpr std::array<PadButton, 4> b2_buttons{
+        PadButton::PAD_BUTTON_START, PadButton::PAD_TRIGGER_Z, PadButton::PAD_TRIGGER_R,
+        PadButton::PAD_TRIGGER_L};
```
Ditto with a comma on the last element.

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,369 @@
+
+        if (payload_size_copy != sizeof(adapter_payload_copy) ||
+            adapter_payload_copy[0] != LIBUSB_DT_HID) {
+            LOG_ERROR(Input, "error reading payload (size: %d, type: %02x)", payload_size_copy,
```
```suggestion
            LOG_ERROR(Input, "error reading payload (size: {}, type: {:02x})", payload_size_copy,
```

We use libfmt, so fmt specifiers should be used instead, otherwise this will crash at runtime if the error case ever occurs.

ameerj replied:
Just for the record, it literally prints the `%d` rather than format it the way I had it 😅

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,369 @@
+    const int get_descriptor_error = libusb_get_device_descriptor(device, &desc);
+    if (get_descriptor_error) {
+        // could not acquire the descriptor, no point in trying to use it.
+        LOG_ERROR(Input, "libusb_get_device_descriptor failed with error: %d",
```
```suggestion
        LOG_ERROR(Input, "libusb_get_device_descriptor failed with error: {}",
```

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,369 @@
+    const int open_error = libusb_open(device, &usb_adapter_handle);
+
+    if (open_error == LIBUSB_ERROR_ACCESS) {
+        LOG_ERROR(Input, "Yuzu can not gain access to this device: ID %04X:%04X.", desc.idVendor,
```
```suggestion
        LOG_ERROR(Input, "Yuzu can not gain access to this device: ID {:04X}:{:04X}.", desc.idVendor,
```

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,369 @@
+        return false;
+    }
+    if (open_error) {
+        LOG_ERROR(Input, "libusb_open failed to open device with error = %d", open_error);
```
```suggestion
        LOG_ERROR(Input, "libusb_open failed to open device with error = {}", open_error);
```

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,369 @@
+    if (kernel_driver_error == 1) {
+        kernel_driver_error = libusb_detach_kernel_driver(usb_adapter_handle, 0);
+        if (kernel_driver_error != 0 && kernel_driver_error != LIBUSB_ERROR_NOT_SUPPORTED) {
+            LOG_ERROR(Input, "libusb_detach_kernel_driver failed with error = %d",
```
```suggestion
            LOG_ERROR(Input, "libusb_detach_kernel_driver failed with error = {}",
```

lioncash commented on src/input_common/gcadapter/gc_adapter.cpp:
```diff
@@ -0,0 +1,369 @@
+
+    const int interface_claim_error = libusb_claim_interface(usb_adapter_handle, 0);
+    if (interface_claim_error) {
+        LOG_ERROR(Input, "libusb_claim_interface failed with error = %d", interface_claim_error);
```
```suggestion
        LOG_ERROR(Input, "libusb_claim_interface failed with error = {}", interface_claim_error);
```

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,163 @@
+};
+
+/// Used to loop through the and assign button in poller
+static constexpr std::array<PadButton, 12> PadButtonArray{
```
This should probably be defined within a cpp file somewhere and be made extern const, otherwise this will be copied into every translation unit that makes use of it (which is two files).

lioncash commented on src/input_common/gcadapter/gc_adapter.h:
```diff
@@ -0,0 +1,163 @@
+    PadButton::PAD_BUTTON_LEFT, PadButton::PAD_BUTTON_RIGHT, PadButton::PAD_BUTTON_DOWN,
+    PadButton::PAD_BUTTON_UP,   PadButton::PAD_TRIGGER_Z,    PadButton::PAD_TRIGGER_R,
+    PadButton::PAD_TRIGGER_L,   PadButton::PAD_BUTTON_A,     PadButton::PAD_BUTTON_B,
+    PadButton::PAD_BUTTON_X,    PadButton::PAD_BUTTON_Y,     PadButton::PAD_BUTTON_START};
```
Comma after last element.
